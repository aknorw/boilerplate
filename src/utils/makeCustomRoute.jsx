import React from 'react'
import PropTypes from 'prop-types'
import { Route } from 'react-router-dom'

const CustomRoute = ({ path, component, ...rest }, prefix = null) => {
  const realPath = prefix ? prefix + path : path
  const route = (
    <Route
      key={realPath}
      path={realPath}
      component={component}
    />
  )
  if (rest.routes) {
    return rest.routes.map(r => CustomRoute(r, path))
  }
  return route
}

CustomRoute.propTypes = {
  path: PropTypes.string.isRequired,
  component: PropTypes.node.isRequired,
}

export default CustomRoute
