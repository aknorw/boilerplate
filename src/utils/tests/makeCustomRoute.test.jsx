import React from 'react'
import { shallow } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'

import makeCustomRoute from '../makeCustomRoute'

describe('Utils - makeCustomRoute', () => {
  it('should be defined', () => {
    expect(makeCustomRoute).toBeDefined()
  })
  it('should render correctly flat route', () => {
    const tree = shallow(
      <MemoryRouter>
        {makeCustomRoute({})}
      </MemoryRouter>,
    )
    expect(tree).toMatchSnapshot()
  })
  it('should render correctly nested route', () => {
    const route = {
      path: '/',
      routes: [{
        path: 'test/',
      }],
    }
    const tree = shallow(
      <MemoryRouter>
        {makeCustomRoute(route)}
      </MemoryRouter>,
    )
    expect(tree).toMatchSnapshot()
  })
})
