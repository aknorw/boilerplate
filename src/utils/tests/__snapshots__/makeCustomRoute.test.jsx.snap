// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils - makeCustomRoute should render correctly flat route 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "pvfpk9",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "pvfpk9",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Route />
</Router>
`;

exports[`Utils - makeCustomRoute should render correctly nested route 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "8w9afm",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "8w9afm",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Route
    key="/test/"
    path="/test/"
  />
</Router>
`;
